### PR TITLE
Add MariaDB test harnesses

### PR DESCRIPTION
**Pull request by christianversloot**
## Describe changes
We are running ZenML on MariaDB instead of MySQL. This afternoon, you (esp @stefannica) helped me fix the upgrade to 0.52.0 by fixing a (seemingly unsupported) database migration to MariaDB. As MariaDB is an open source fork of (some version of) MySQL that is used quite frequently, we've concluded this afternoon that it's to be included in ZenML tests. By consequence, as suggested by Stefan this afternoon, this PR serves as (an initial attempt to) add MariaDB test harnesses into ZenML :) 

Currently, it does the following:
* Rename the MySQL `local_docker` and `server_docker_compose` test harnesses into `*_mysql` ones, allowing for support of more database engines, also renaming them in `zenml/tests/harness/deployment/__init__.py` so that they remain used.
* Add a `local_docker_mariadb` variant, which clones the MySQL variant, but which uses the most recent MariaDB minor release: `mariadb:10.6`.
* Add a `server_docker_compose_mariadb` variant, which clones the MySQL variant, but which uses the most recent MariaDB minor release: `mariadb:10.6`. 
* Add the new variants to `__all__` in the deployment tests `zenml.tests.harness.deployment.__init__` file/module.

A few questions that are on my mind before we can merge the PR:
* I haven't run the tests locally yet. Stefan sent me the [testing document](https://github.com/zenml-io/zenml/blob/main/tests/README.md) this afternoon. Do you want me to run the tests locally or are they automatically picked up by GitHub CI/CD testing?
* What should be the minimum MariaDB minor release that we support? 10.6 seems to be the current LTS version (and our AWS RDS database runs 10.6.x), so that seems to be an adequate minimum version?
* Are there things that I'm missing? I've cloned the `local_docker` and `server_docker_compose` (MySQL) variants and altered them so that they start up a Docker container with the expected values such as `MARIADB_ROOT_PASSWORD` (instead of `MYSQL_DEFAULT_PASSWORD`); also, I've added them to `zenml.tests.harness.deployment.__init__`, but is there more to be done?

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by stefannica — CHANGES_REQUESTED** [review]
Thank you so much for the fast submission ! This is almost ready to merge, it's just missing a few small pieces to be fully functional.

> I haven't run the tests locally yet. Stefan sent me the [testing document](https://github.com/zenml-io/zenml/blob/main/tests/README.md) this afternoon. Do you want me to run the tests locally or are they automatically picked up by GitHub CI/CD testing?

You'll have add a few more things to use the mariadb environment in the CI:

* configure mariadb based deployments and environments in the `tests/harness/cfg` folder
* add the new environments as options in `.github/workflows/integration-test.yaml`
* add the new environment to the test matrix in `.github/workflows/ci.yaml`

It would be best to run at least a couple of tests locally once you set up the environments, e.g.:

```
./zen-test environment provision docker-compose-server-mariadb
pytest tests/integration/functional/zen_stores/test_zen_store.py --environment docker-compose-server-mariadb --no-provision
```

> What should be the minimum MariaDB minor release that we support? 10.6 seems to be the current LTS version (and our AWS RDS database runs 10.6.x), so that seems to be an adequate minimum version?

I would leave that up to you, but LTS sounds like a good choice.

> Are there things that I'm missing? I've cloned the local_docker and server_docker_compose (MySQL) variants and altered them so that they start up a Docker container with the expected values such as MARIADB_ROOT_PASSWORD (instead of MYSQL_DEFAULT_PASSWORD); also, I've added them to zenml.tests.harness.deployment.__init__, but is there more to be done?

Yes, a few more things are needed:

* you need unique identifiers for the new deployment types (see my comment)
* you need to configure some mariadb-based deployments and environments in the test harness configuration (under `tests/harness/cfg`)
* you need to update the github workflows with the new environment names

You'll also probably need to rename some of the existing deployments and environments in the harness config to mention mysql instead of just `docker`, which means the new names also need to be reflected in the github workflows.

File: tests/harness/deployment/local_docker_mariadb.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                f"deployment."
+            )
+
+        return DeploymentStoreConfig(
+            url=f"mysql://root:{MARIADB_ROOT_PASSWORD}@127.0.0.1:{port}/zenml"
+        )
+
+
+LocalDockerMariadbTestDeployment.register_deployment_class(
+    type=DeploymentType.LOCAL, setup=DeploymentSetup.DOCKER
```
[the remark this answers is not in the document]
Comment (christianversloot) [reply tests/harness/deployment/local_docker_mariadb.py:211]: I think that is a good idea. Perhaps, we can add `DOCKER_COMPOSE` as an extra enum value to `ServerType`, to distinguish Docker Compose based setups from vanilla Docker setups.

**Comment by christianversloot** [comment]
Thanks for your review! Summarizing into an action plan, I come to this:

1. Resolving the issue with unique identifiers for the new deployment types.
2. Configuring mariadb based deployments and environments in the tests/harness/cfg folder while renaming existing deployments and environments in that folder.
3. Adding the new environments as options in .github/workflows/integration-test.yaml
4. Adding the new environments to the test matrix in .github/workflows/ci.yaml
5. Running some tests locally.

I'll make sure to work on these items this week. Please let me know if I am missing something.

**Comment by stefannica** [comment]
> I'm not sure though why the unit tests are failing... any insights @stefannica? Thanks :)

Yes, the `server_sqlite.py` and `local_sqlite.py` deployments both use the same identifier pair:
```
ServerSqliteTestDeployment.register_deployment_class(
    type=ServerType.LOCAL, setup=DatabaseType.SQLITE
)
```

I imagine this causes some inconsistency in how the deployments are provisioned and then used.

**Comment by stefannica** [comment]
@christianversloot I pushed some changes to your PR to help get it to the finish line faster. I hope that's okay.

**Review by stefannica — APPROVED** [review]
Thanks again for putting this together ! Looking forward to having this run as part of the day-to-day CI.

**Comment by stefannica** [comment]
Just a sneak peak of the new way of organizing deployments. I think this looks great:

```
$ ./zen-test deployment list
┏━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━┯━━━━━━━━━━┯━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┯━━━━━━━━━━┯━━━━━━━━━┯━━━━━┓
┃         NAME          │ SERVER │ DATABASE │ DESCRIPTION                                                                     │ DISABLED │ RUNNING │ URL ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃        default        │ NONE   │ SQLITE   │ Default deployment.                                                             │          │ ✅      │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃     client-mysql      │ NONE   │ MYSQL    │ Local client connected directly to MySQL running in container.                  │          │         │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃    client-mariadb     │ NONE   │ MARIADB  │ Local client connected directly to MariaDB running in container.                │          │         │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃     local-server      │ LOCAL  │ SQLITE   │ Local ZenML server running as daemon process using the default SQLite database. │          │         │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃     docker-server     │ DOCKER │ SQLITE   │ Local ZenML server running in docker using the default SQLite database.         │          │         │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃  docker-server-mysql  │ DOCKER │ MYSQL    │ Local ZenML server and MySQL both running in docker with docker-compose.        │          │         │     ┃
┠───────────────────────┼────────┼──────────┼─────────────────────────────────────────────────────────────────────────────────┼──────────┼─────────┼─────┨
┃ docker-server-mariadb │ DOCKER │ MARIADB  │ Local ZenML server and MariaDB both running in docker with docker-compose.      │          │         │     ┃
┗━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━┷━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━┷━━━━━━━━━┷━━━━━┛

```

**Comment by stefannica** [comment]
@christianversloot this might also be interesting for you: https://github.com/zenml-io/zenml/pull/2170

**Comment by christianversloot** [comment]
@stefannica yes, super nice! Happy to see it's integrated so quickly!

**Pull request merged**